### PR TITLE
Add libparsec/README.md

### PR DESCRIPTION
**Pull request by touilleMan**

### CODE REVIEW

File: libparsec/README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+- `libparsec` (i.e. code in `./src`): crate exposed to the outer world, the JS bindings (i.e. `../bindings`)
+  exposed the content of this crate.
+- `libparsec_*` (i.e. code in `./crates/`): internal crates, `libparsec` itself mostly
+  re-expose them.
+
+> Notes:
+>
+> - Currently the CLI relies on the internal `libparsec_*` crates instead of the
+>   `libparsec`. This is for historical reasons and should eventualy be fixed.
+> - For simplicity, internal crates don't have there `libparsec_` suffix in the
```
Comment (AureliaDolo) [inline libparsec/README.md:14]:
```suggestion
> - For simplicity, internal crates don't have their `libparsec_` suffix in the
```

File: libparsec/README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+> Notes:
+>
+> - `serialization_format` also provides a macro to generate Python bindings
+>   for the protocol (used by the server). Note this code is beyond mere mortal
+>   comprehension, debugging requires sacrificing your soul to Satan.
+
+### Platform crates
+
+Crates with platform in there name have different implementations depending on
```
Comment (AureliaDolo) [inline libparsec/README.md:61]:
```suggestion
Crates with platform in their name have different implementations depending on
```

**Review by AureliaDolo — APPROVED** [review]
Thanks ! :sparkles:

File: libparsec/README.md (hunk 2 above)
Comment (touilleMan) [reply libparsec/README.md]: done.

File: libparsec/README.md (hunk 1 above)
Comment (touilleMan) [reply libparsec/README.md]: done.

**Review by mmmarcos — COMMENTED** [review]
We should add the following line at the beginning of `libparsec/src/lib.rs`:
```rust
#![doc = include_str!("../README.md")]
```

This way, it would be included by `cargo doc` ✨ 
image

File: libparsec/README.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+### Why most tests are in `tests/unit` ?
+
+#### TL;DR
+
+Stick to the following rule:
+
+- All tests are unit tests.
+- All test code must be in `tests/unit` (this works by using `#[path = "../../tests/unit/my_test.rs"]`
+  to overwrite the test module location).
```
Comment (AureliaDolo) [inline libparsec/README.md:125]:
```suggestion
- All test code must be in `tests/unit` (this works by using `#[path = "../../tests/unit/my_test.rs"]`
  to overwrite the test module location in the tested crate, where you'd put your usual unit tests).
```

I'm not sure about the exact formulation, but what I'm trying to say is to explicit where to add the `#[path = "..."]` line

**Review by AureliaDolo — APPROVED** [review]

File: libparsec/README.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
+# Welcome to libparsec \o/
+
+## Crates layout
+
+- `libparsec` (i.e. code in `./src`): crate exposed to the outer world, the JS bindings (i.e. `../bindings`)
+  exposed the content of this crate.
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:6]: :+1: done.

File: libparsec/README.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  re-expose them.
+
+> Notes:
+>
+> - Currently the CLI relies on the internal `libparsec_*` crates instead of the
+>   `libparsec`. This is for historical reasons and should eventually be fixed.
+> - For simplicity, internal crates don't have their `libparsec_` suffix in the
+>   folder hierarchy (e.g. `libparsec_types` is located at `./crates/types`).
+
+Among internal crates, there is:
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:17]: done.

File: libparsec/README.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+### Client
+
+The client is the component that runs a device.
+It is currently divided into two crates:
+
+- `client_connection`: Handle connection to the server.
+- `client`: Contains everything else: invitation&enrollment, certificate&manifest validation,
+  workspace operations, monitors running background tasks etc.
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:26]: done.

File: libparsec/README.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+> Notes:
+>
+> The `client` crate can be seen as the "main" crate that implement most of the
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:30]: done.

File: libparsec/README.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+> Notes:
+>
+> - Currently the CLI relies on the internal `libparsec_*` crates instead of the
+>   `libparsec`. This is for historical reasons and should eventually be fixed.
+> - For simplicity, internal crates don't have their `libparsec_` suffix in the
+>   folder hierarchy (e.g. `libparsec_types` is located at `./crates/types`).
+
+Among internal crates, there is:
+
+### Client
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:19]: done.

File: libparsec/README.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+- `client_connection`: Handle connection to the server.
+- `client`: Contains everything else: invitation&enrollment, certificate&manifest validation,
+  workspace operations, monitors running background tasks etc.
+
+> Notes:
+>
+> The `client` crate can be seen as the "main" crate that implement most of the
+> functional logic of the application.
+
+### Base types
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:33]: done.

File: libparsec/README.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+The base types can be divided into three:
+
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:36]: The next line "Communication between client and server is done trough an API that is" was an error, so I removed it instead

File: libparsec/README.md [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+   We call this simply the "data" and it is defined in `types/schema`.
+3. What is used only internally (i.e. in-memory and never serialized).
+
+The key point here is backward/forward compatibility as serialized data are
+produced and consumed by different version of Parsec.
+
+- `protocol`: Implement the protocol, and the API version.
+- `types`: Implement the data, and the types that are only for internal use.
+- `serialization_format`: Define the format used for serialization (for both
+  protocol and data).
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:51]: done.

File: libparsec/README.md [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+### Platform crates
+
+Crates with platform in their name have different implementations depending on
+the platform they are compiled for.
+
+For instance `libparsec_platform_mountpoint` uses FUSE on UNIX, WinFSP on Windows,
+and is not compatible with web.
+
+The Corollary is that crates without `platform` in their name are platform-independent,
+and hence can only be tested on a single platform.
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:68]: done.

File: libparsec/README.md [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+
+The platform crates:
+
+- `platform_async`: Asynchronous programming works differently between web and native
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:72]: done.

File: libparsec/README.md [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+  (the former has to integrate with the javascript runtime, the latter uses Tokio), so
+  this crate exposes platform-agnostic utilities (e.g. spawn a future, async lock etc.).
+- `platform_device_loader`: Load the device keys file from disk.
+- `platform_http_proxy`: Handle HTTP proxy.
+- `platform_ipc`: Allow communication between processes to ensure a device is not
+                  concurrently used.
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:78]: done.

File: libparsec/README.md [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+>
+> - From an external point of view, the `tests_fixture` re-expose everything needed.
+>   So when writing tests, `use libparsec_tests_fixtures::prelude::*` is the single-no-brainer-one-liner™.
+> - `tests_fixtures`'s name is mostly an historical artifact, as it contains far
+>   more than only fixtures !
+
+### Special crates
+
+- `crypto` Implement all the crypto algo (signing, encryption, hashing etc.). There is two
+  implementation for those: Rustcrypto (pure Rust) and Sodiumoxide (based on libsodium).
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:112]: done.

File: libparsec/README.md [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+
+In Rust there is concept of unit vs integration testing:
+
+- Unit test are within the tested crate.
+- Integration are code outside the tested crate.
+
+> « But I don't care about your "crate" concept ! Defining a test as unit or integration
+> should depend of what it does, not where it comes from ! »
+
+Well tough luck ! That's how Rust works, so you'd better play along ¯\\_(ツ)_/¯
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:138]: done. (I've used `“crate”` for quote-in-quote

File: libparsec/README.md [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+  check (if the check access internals).
+- It's very common to access internals when testing (e.g. define a dump method for test
+  only that returns an internal type).
+- Unit tests code is located in the `src` directory and might even be inlined with
+  the code to test. This add a lot of noise when reading the code :(
+
+So to solve this we stick to the "all test code is unit test located in `tests/unit` folder".
+
+The drawback to this approach is it [increases compilation time](https://xkcd.com/303/)
+(because the crate is the unit of compilation in Rust, remember ? ^^) since 1) all
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/README.md:159]: done.

File: libparsec/README.md (hunk 3 above)
Comment (touilleMan) [reply libparsec/README.md:125]: I think we don't need extensive explanation here given this pattern is already used extensively in the code (so it's easy to just search for `#[path = ` in the codebase to understand how it is used).

**Review by mmmarcos — APPROVED** [review]
:100:

**Pull request merged**